File: user_tools/connection_examples/01_Query_AKVEG_Data.py
# -*- coding: utf-8 -*-
# ---------------------------------------------------------------------------
# Query data from AKVEG Database
# Author: Timm Nawrocki, Amanda Droghini, Alaska Center for Conservation Science
# Last Updated: 2025-05-06
# Usage: Script should be executed in Python 3.12+. Requires psycopg2.
# Description: "Compile data views" provides an example of compiling a set of data views for a user-specified region from the AKVEG Database.
# ---------------------------------------------------------------------------

# Import packages
import os
import pandas as pd
import geopandas as gpd
from akutils import connect_database_postgresql  # Optional (see below)
from akutils import query_to_dataframe  # Optional (see below)

# IF NOT USING AKUTILS, UNCOMMENT LINES BELOW (FOR IDE THAT AUTO-RECOGNIZES INIT FILE)
# from pull_functions import connect_database_postgresql
# from pull_functions import query_to_dataframe

#### SET UP DIRECTORIES AND FILES
####------------------------------

# Set root directory (modify to your folder structure)
drive = 'C:/'
root_folder = 'ACCS_Work'

# Define input folders (modify to your folder structure)
database_repository = os.path.join(drive, root_folder, 'Repositories/akveg-database')
credentials_folder = os.path.join(drive, root_folder, 'Example/Credentials/akveg_public_read')
input_folder = os.path.join(drive, root_folder, 'Example/Data_Input')
output_folder = os.path.join(input_folder, 'plot_data')

# Define input files
domain_input = os.path.join(input_folder, 'region_data/AlaskaYukon_ProjectDomain_v2.0_3338.shp')
region_input = os.path.join(input_folder, 'region_data/AlaskaYukon_Regions_v2.0_3338.shp')
fireyear_input = os.path.join(input_folder, 'ancillary_data/AlaskaYukon_FireYear_10m_3338.tif')

# Define output files
taxa_output = os.path.join(output_folder, '00_taxonomy.csv')
project_output = os.path.join(output_folder, '01_project.csv')
site_visit_output = os.path.join(output_folder, '03_site_visit.csv')
site_point_output = os.path.join(output_folder, '03_site_point_3338.shp')
vegetation_output = os.path.join(output_folder, '05_vegetation.csv')
abiotic_output = os.path.join(output_folder, '06_abiotic_top_cover.csv')
tussock_output = os.path.join(output_folder, '07_whole_tussock_cover.csv')
ground_output = os.path.join(output_folder, '08_ground_cover.csv')
structural_output = os.path.join(output_folder, '09_structural_group_cover.csv')
shrub_output = os.path.join(output_folder, '11_shrub_structure.csv')
environment_output = os.path.join(output_folder, '12_environment.csv')
soilmetrics_output = os.path.join(output_folder, '13_soil_metrics.csv')
soilhorizons_output = os.path.join(output_folder, '14_soil_horizons.csv')

# Define queries
taxa_file = os.path.join(database_repository, 'queries/00_taxonomy.sql')
project_file = os.path.join(database_repository, 'queries/01_project.sql')
site_visit_file = os.path.join(database_repository, 'queries/03_site_visit.sql')
vegetation_file = os.path.join(database_repository, 'queries/05_vegetation.sql')
abiotic_file = os.path.join(database_repository, 'queries/06_abiotic_top_cover.sql')
tussock_file = os.path.join(database_repository, 'queries/07_whole_tussock_cover.sql')
ground_file = os.path.join(database_repository, 'queries/08_ground_cover.sql')
structural_file = os.path.join(database_repository, 'queries/09_structural_group_cover.sql')
shrub_file = os.path.join(database_repository, 'queries/11_shrub_structure.sql')
environment_file = os.path.join(database_repository, 'queries/12_environment.sql')
soilmetrics_file = os.path.join(database_repository, 'queries/13_soil_metrics.sql')
soilhorizons_file = os.path.join(database_repository, 'queries/14_soil_horizons.sql')

# Read local data
domain_shape = gpd.read_file(domain_input)
region_shape = gpd.read_file(region_input)

# Get geometry for intersection (example to subset data by Boreal)
#intersect_geometry = region_shape[region_shape['region'].isin(
#    ['Alaska-Yukon Southern',
#     'Alaska-Yukon Central',
#     'Alaska-Yukon Northern',
#     'Alaska Western',
#     'Alaska Southwest']
#)]

# Get geometry for intersection (example to subset data by Arctic)
intersect_geometry = region_shape[region_shape['region'].isin(
    ['Arctic Northern',
     'Arctic Western']
)]

#### QUERY AKVEG DATABASE
####------------------------------

# Create a connection to the AKVEG PostgreSQL database
authentication_file = os.path.join(credentials_folder, 'authentication_akveg_public_read.csv')
database_connection = connect_database_postgresql(authentication_file)

# Read taxonomy standard from AKVEG Database
taxa_read = open(taxa_file, 'r')
taxa_query = taxa_read.read()
taxa_read.close()
taxa_data = query_to_dataframe(database_connection, taxa_query)

# Read site visit data from AKVEG Database
site_visit_read = open(site_visit_file, 'r')
site_visit_query = site_visit_read.read()
site_visit_read.close()
site_visit_data = query_to_dataframe(database_connection, site_visit_query)
site_visit_data['obs_datetime'] = pd.to_datetime(site_visit_data['observe_date'])
site_visit_data['obs_year'] = site_visit_data['obs_datetime'].dt.year

# Create geodataframe
site_visit_data = gpd.GeoDataFrame(
    site_visit_data,
    geometry=gpd.points_from_xy(site_visit_data.longitude_dd,
                                site_visit_data.latitude_dd),
    crs='EPSG:4269')

# Convert geodataframe to EPSG:3338
site_visit_data = site_visit_data.to_crs(crs='EPSG:3338')

# Extract coordinates in EPSG:3338
site_visit_data['cent_x'] = site_visit_data.geometry.x
site_visit_data['cent_y'] = site_visit_data.geometry.y

# Subset points to map domain (example to subset using a feature class)
site_visit_data = gpd.clip(site_visit_data, domain_shape)

# Subset points to those within the target zone (example to subset using a feature class selection)
site_visit_data = gpd.clip(site_visit_data, intersect_geometry)

# Example filter by project code (uncomment line below)
# site_visit_data = site_visit_data[site_visit_data['project_code'] == 'accs_nelchina_2023']

# Example filter by observation year (uncomment line below)
# site_visit_data = site_visit_data[site_visit_data['obs_year'] >= 2000]

# Example filter by perspective (uncomment line below)
# site_visit_data = site_visit_data[site_visit_data['perspective'] == 'ground']

# Select columns
site_visit_data = site_visit_data[['site_visit_code', 'project_code', 'site_code', 'data_tier',
                                   'observe_date', 'scope_vascular', 'scope_bryophyte', 'scope_lichen',
                                   'perspective', 'cover_method', 'structural_class', 'homogeneous',
                                   'plot_dimensions_m', 'latitude_dd', 'longitude_dd',
                                   'cent_x', 'cent_y']]

# Rename fields for export as shapefile to meet shapefile field character length constraint
export_point_data = site_visit_data.rename(columns={'site_visit_code': 'st_vst',
                                                    'project_code': 'prjct_cd',
                                                    'site_code': 'st_code',
                                                    'observe_date': 'obs_date',
                                                    'scope_vascular': 'scp_vasc',
                                                    'scope_bryophyte': 'scp_bryo',
                                                    'scope_lichen': 'scp_lich',
                                                    'perspective': 'perspect',
                                                    'cover_method': 'cvr_mthd',
                                                    'structural_class': 'strc_class',
                                                    'homogeneous': 'hmgneous',
                                                    'plot_dimensions_m': 'plt_dim_m',
                                                    'latitude_dd': 'lat_dd',
                                                    'longitude_dd': 'long_dd'
                                                    })

# Export site visit data to shapefile
site_point_data = gpd.GeoDataFrame(
    export_point_data,
    geometry=gpd.points_from_xy(site_visit_data.cent_x,
                                site_visit_data.cent_y),
    crs='EPSG:3338')
site_point_data.to_file(site_point_output)  # Optional to check point selection in a GIS

# Write where statement for site visits
input_sql = '\r\nWHERE site_visit.site_visit_code IN ('
for site_visit in site_visit_data['site_visit_code']:
    input_sql = input_sql + r"'" + site_visit + r"', "
input_sql = input_sql[:-2] + r');'

# Read project data from AKVEG Database for selected site visits
project_read = open(project_file, 'r')
project_query = project_read.read()
project_read.close()
project_query = project_query.replace(';', input_sql)
project_data = query_to_dataframe(database_connection, project_query).sort_values('project_code')

# Read vegetation cover data from AKVEG Database for selected site visits
vegetation_read = open(vegetation_file, 'r')
vegetation_query = vegetation_read.read()
vegetation_read.close()
vegetation_query = vegetation_query.replace(';', input_sql)
vegetation_data = query_to_dataframe(database_connection, vegetation_query)

# Read abiotic top cover data from AKVEG Database for selected site visits
abiotic_read = open(abiotic_file, 'r')
abiotic_query = abiotic_read.read()
abiotic_read.close()
abiotic_query = abiotic_query.replace(';', input_sql)
abiotic_data = query_to_dataframe(database_connection, abiotic_query)

# Read whole tussock cover data from AKVEG Database for selected site visits
tussock_read = open(tussock_file, 'r')
tussock_query = tussock_read.read()
tussock_read.close()
tussock_query = tussock_query.replace(';', input_sql)
tussock_data = query_to_dataframe(database_connection, tussock_query)

# Read ground cover data from AKVEG Database for selected site visits
ground_read = open(ground_file, 'r')
ground_query = ground_read.read()
ground_read.close()
ground_query = ground_query.replace(';', input_sql)
ground_data = query_to_dataframe(database_connection, ground_query)

# Read structural group cover data from AKVEG Database for selected site visits
structural_read = open(structural_file, 'r')
structural_query = structural_read.read()
structural_read.close()
structural_query = structural_query.replace(';', input_sql)
structural_data = query_to_dataframe(database_connection, structural_query)

# Read shrub structure data from AKVEG Database for selected site visits
shrub_read = open(shrub_file, 'r')
shrub_query = shrub_read.read()
shrub_read.close()
shrub_query = shrub_query.replace(';', input_sql)
shrub_data = query_to_dataframe(database_connection, shrub_query)

# Read environment data from AKVEG Database for selected site visits
environment_read = open(environment_file, 'r')
environment_query = environment_read.read()
environment_read.close()
environment_query = environment_query.replace(';', input_sql)
environment_data = query_to_dataframe(database_connection, environment_query)

# Read soil metrics data from AKVEG Database for selected site visits
soilmetrics_read = open(soilmetrics_file, 'r')
soilmetrics_query = soilmetrics_read.read()
soilmetrics_read.close()
soilmetrics_query = soilmetrics_query.replace(';', input_sql)
soilmetrics_data = query_to_dataframe(database_connection, soilmetrics_query)

# Read soil horizons data from AKVEG Database for selected site visits
soilhorizons_read = open(soilhorizons_file, 'r')
soilhorizons_query = soilhorizons_read.read()
soilhorizons_read.close()
soilhorizons_query = soilhorizons_query.replace(';', input_sql)
soilhorizons_data = query_to_dataframe(database_connection, soilhorizons_query)

# Check number of cover observations per project
project_check = pd.merge(vegetation_data, site_visit_data, on='site_visit_code', how='left')[['project_code',
                                                                                              'site_visit_code']]
project_check = project_check.groupby(['project_code']).count().rename(columns={'site_visit_code': 'obs_n'})
project_check['project_code'] = project_check.index
project_check = project_check.reset_index(drop=True)[['project_code', 'obs_n']]

# Export data to csv files
taxa_data.to_csv(taxa_output, index=False, encoding='utf-8')
project_data.to_csv(project_output, index=False, encoding='utf-8')
site_visit_data.to_csv(site_visit_output, index=False, encoding='utf-8')
vegetation_data.to_csv(vegetation_output, index=False, encoding='utf-8')
abiotic_data.to_csv(abiotic_output, index=False, encoding='utf-8')
tussock_data.to_csv(tussock_output, index=False, encoding='utf-8')
ground_data.to_csv(ground_output, index=False, encoding='utf-8')
structural_data.to_csv(structural_output, index=False, encoding='utf-8')
shrub_data.to_csv(shrub_output, index=False, encoding='utf-8')
environment_data.to_csv(environment_output, index=False, encoding='utf-8')
soilmetrics_data.to_csv(soilmetrics_output, index=False, encoding='utf-8')
soilhorizons_data.to_csv(soilhorizons_output, index=False, encoding='utf-8')

File: user_tools/connection_examples/01_Query_AKVEG_Data.R
# -*- coding: utf-8 -*-
# ---------------------------------------------------------------------------
# Query data from AKVEG Database
# Author: Timm Nawrocki, Amanda Droghini, Alaska Center for Conservation Science
# Last Updated: 2025-05-06
# Usage: Script should be executed in R 4.4.3+.
# Description: "Query data from AKVEG Database" is an example script to pull data from the AKVEG Database for all available, non-metadata tables. The script connects to the AKVEG database, executes queries, and performs simple spatial analyses (i.e., subset the data to specific study areas, extract raster values to surveyed plots). The outputs are a series of CSV files (one for each non-metadata table in the database) whose results are restricted to the study area in the script.
# ---------------------------------------------------------------------------

# Import required libraries ----
library(dplyr)
library(fs)
library(janitor)
library(lubridate)
library(readr)
library(readxl)
library(RPostgres)
library(sf)
library(stringr)
library(terra)
library(tibble)
library(tidyr)

#### Set up directories and files ------------------------------

# Set root directory (modify to your folder structure)
drive = 'C:'
root_folder = 'ACCS_Work'

# Define input folders (modify to your folder structure)
database_repository = path(drive, root_folder, 'Repositories/akveg-database-public')
credentials_folder = path(drive, root_folder, 'Example/Credentials/akveg_public_read')
input_folder = path(drive, root_folder, 'Example/Data_Input')
output_folder = path(drive, root_folder, 'Example/Data_Input', 'plot_data')

# Define input files
domain_input = path(input_folder, 'region_data/AlaskaYukon_ProjectDomain_v2.0_3338.shp')
region_input = path(input_folder, 'region_data/AlaskaYukon_Regions_v2.0_3338.shp')
fireyear_input = path(input_folder, 'ancillary_data/AlaskaYukon_FireYear_10m_3338.tif')

# Define output files
taxa_output = path(output_folder, '00_taxonomy.csv')
project_output = path(output_folder, '01_project.csv')
site_visit_output = path(output_folder, '03_site_visit.csv')
site_point_output = path(output_folder, '03_site_point_3338.shp')
vegetation_output = path(output_folder, '05_vegetation.csv')
abiotic_output = path(output_folder, '06_abiotic_top_cover.csv')
tussock_output = path(output_folder, '07_whole_tussock_cover.csv')
ground_output = path(output_folder, '08_ground_cover.csv')
structural_output = path(output_folder, '09_structural_group_cover.csv')
shrub_output = path(output_folder, '11_shrub_structure.csv')
environment_output = path(output_folder, '12_environment.csv')
soilmetrics_output = path(output_folder, '13_soil_metrics.csv')
soilhorizons_output = path(output_folder, '14_soil_horizons.csv')

# Define queries
taxa_file = path(database_repository, 'queries/00_taxonomy.sql')
project_file = path(database_repository, 'queries/01_project.sql')
site_visit_file = path(database_repository, 'queries/03_site_visit.sql')
vegetation_file = path(database_repository, 'queries/05_vegetation.sql')
abiotic_file = path(database_repository, 'queries/06_abiotic_top_cover.sql')
tussock_file = path(database_repository, 'queries/07_whole_tussock_cover.sql')
ground_file = path(database_repository, 'queries/08_ground_cover.sql')
structural_file = path(database_repository, 'queries/09_structural_group_cover.sql')
shrub_file = path(database_repository, 'queries/11_shrub_structure.sql')
environment_file = path(database_repository, 'queries/12_environment.sql')
soilmetrics_file = path(database_repository, 'queries/13_soil_metrics.sql')
soilhorizons_file = path(database_repository, 'queries/14_soil_horizons.sql')

# Read local data ----
domain_shape = st_read(domain_input)
region_shape = st_read(region_input)
fireyear_raster = rast(fireyear_input)

# Get geometry for intersection (example to subset data by Boreal)
#intersect_geometry = st_geometry(region_shape[region_shape$region == 'Alaska-Yukon Southern'
#                                              | region_shape$region == 'Alaska-Yukon Central'
#                                              | region_shape$region == 'Alaska-Yukon Northern'
#                                              | region_shape$region == 'Alaska Western'
#                                              | region_shape$region == 'Alaska Southwest',])

# Get geometry for intersection (example to subset data by Arctic)
intersect_geometry = st_geometry(region_shape[region_shape$region == 'Arctic Northern'
                                              | region_shape$region == 'Arctic Western',])

#### Query AKVEG database ------------------------------

# Import database connection function
connection_script = path(database_repository, 'pull_functions', 'connect_database_postgresql.R')
source(connection_script)

# Create a connection to the AKVEG PostgreSQL database
authentication = path(credentials_folder, 'authentication_akveg_public_read.csv')
database_connection = connect_database_postgresql(authentication)

# Read taxonomy standard from AKVEG Database
taxa_query = read_file(taxa_file)
taxa_data = as_tibble(dbGetQuery(database_connection, taxa_query))

# Read site visit data from AKVEG Database
site_visit_query = read_file(site_visit_file)
site_visit_data = as_tibble(dbGetQuery(database_connection, site_visit_query)) %>%
  # Convert geometries to points with EPSG:4269
  st_as_sf(x = ., coords = c('longitude_dd', 'latitude_dd'), crs = 4269, remove = FALSE) %>%
  # Reproject coordinates to EPSG 3338
  st_transform(crs = st_crs(3338)) %>%
  # Add EPSG:3338 centroid coordinates
  mutate(cent_x = st_coordinates(.$geometry)[,1],
         cent_y = st_coordinates(.$geometry)[,2]) %>%
  # Subset points to map domain (example to subset using a feature class)
  st_intersection(st_geometry(domain_shape)) %>%
  # Subset points to those within the target zone (example to subset using a feature class selection)
  st_intersection(intersect_geometry) %>%
  # Extract raster data to points
  mutate(fire_year = terra::extract(fireyear_raster, ., raw=TRUE)[,2]) %>%
  # Drop geometry
  st_zm(drop = TRUE, what = "ZM") %>%
  # Example filter by project code (uncomment line below)
  #filter(project_code == 'accs_nelchina_2023') %>%
  # Example filter by observation year (uncomment line below)
  #filter(year(observe_date) >= 2000) %>%
  # Example filter by perspective (uncomment line below)
  #filter(perspective == 'ground') %>%
  # Select columns
  dplyr::select(site_visit_code, project_code, site_code, data_tier, observe_date, scope_vascular, scope_bryophyte, scope_lichen,
                perspective, cover_method, structural_class, fire_year, homogeneous, plot_dimensions_m,
                latitude_dd, longitude_dd, cent_x, cent_y, geometry)

# Export site visit data to shapefile
site_visit_data %>%
  # Rename fields so that they are within the character length limits
  rename(st_vst = site_visit_code,
         prjct_cd = project_code,
         st_code = site_code,
         obs_date = observe_date,
         scp_vasc = scope_vascular,
         scp_bryo = scope_bryophyte,
         scp_lich = scope_lichen,
         perspect = perspective,
         cvr_mthd = cover_method,
         strc_class = structural_class,
         hmgneous = homogeneous,
         plt_dim_m = plot_dimensions_m,
         lat_dd = latitude_dd,
         long_dd = longitude_dd) %>%
  st_write(site_point_output, append = FALSE) # Optional to check point selection in a GIS

# Write where statement for site visits to apply site visit codes obtained in the spatial intersection above to the SQL queries to restrict data from other tables to only those sites that are within the area of interest 
input_sql = site_visit_data %>%
  # Drop geometry
  st_drop_geometry() %>%
  select(site_visit_code) %>%
  # Format site visit codes
  mutate(site_visit_code = paste('\'', site_visit_code, '\'', sep = '')) %>%
  # Collapse rows
  summarize(site_visit_code = paste(site_visit_code, collapse=", ")) %>%
  # Pull result out of dataframe
  pull(site_visit_code)
where_statement = paste('\r\nWHERE site_visit.site_visit_code IN (',
                        input_sql,
                        ');',
                        sep = '')

# Read project data from AKVEG Database for selected site visits
project_query = read_file(project_file) %>%
  # Modify query with where statement
  str_replace(., ';', where_statement)
project_data = as_tibble(dbGetQuery(database_connection, project_query)) %>%
  arrange(project_code)

# Read vegetation cover data from AKVEG Database for selected site visits
vegetation_query = read_file(vegetation_file) %>%
  # Modify query with where statement
  str_replace(., ';', where_statement)
vegetation_data = as_tibble(dbGetQuery(database_connection, vegetation_query))

# Read abiotic top cover data from AKVEG Database for selected site visits
abiotic_query = read_file(abiotic_file) %>%
  # Modify query with where statement
  str_replace(., ';', where_statement)
abiotic_data = as_tibble(dbGetQuery(database_connection, abiotic_query))

# Read whole tussock cover data from AKVEG Database for selected site visits
tussock_query = read_file(tussock_file) %>%
  # Modify query with where statement
  str_replace(., ';', where_statement)
tussock_data = as_tibble(dbGetQuery(database_connection, tussock_query))

# Read ground cover data from AKVEG Database for selected site visits
ground_query = read_file(ground_file) %>%
  # Modify query with where statement
  str_replace(., ';', where_statement)
ground_data = as_tibble(dbGetQuery(database_connection, ground_query))

# Read structural group cover data from AKVEG Database for selected site visits
structural_query = read_file(structural_file) %>%
  # Modify query with where statement
  str_replace(., ';', where_statement)
structural_data = as_tibble(dbGetQuery(database_connection, structural_query))

# Read shrub structure data from AKVEG Database for selected site visits
shrub_query = read_file(shrub_file) %>%
  # Modify query with where statement
  str_replace(., ';', where_statement)
shrub_data = as_tibble(dbGetQuery(database_connection, shrub_query))

# Read environment data from AKVEG Database for selected site visits
environment_query = read_file(environment_file) %>%
  # Modify query with where statement
  str_replace(., ';', where_statement)
environment_data = as_tibble(dbGetQuery(database_connection, environment_query))

# Read soil metrics data from AKVEG Database for selected site visits
soilmetrics_query = read_file(soilmetrics_file) %>%
  str_replace(., ';', where_statement)
soilmetrics_data = as_tibble(dbGetQuery(database_connection, soilmetrics_query))

# Read soil horizons data from AKVEG Database for selected site visits
soilhorizons_query = read_file(soilhorizons_file) %>%
  str_replace(., ';', where_statement)
soilhorizons_data = as_tibble(dbGetQuery(database_connection, soilhorizons_query))

# Check number of cover observations per project
project_check = vegetation_data %>%
  left_join(site_visit_data, join_by('site_visit_code')) %>%
  group_by(project_code) %>%
  summarize(obs_n = n())

# Export data to csv files ----
taxa_data %>%
  write_csv(., file = taxa_output)
project_data %>%
  write_csv(., file = project_output)
site_visit_data %>%
  st_drop_geometry() %>%
  write_csv(., file = site_visit_output)
vegetation_data %>%
  write_csv(., file = vegetation_output)
abiotic_data %>%
  write_csv(., file = abiotic_output)
tussock_data %>%
  write_csv(., file = tussock_output)
ground_data %>%
  write_csv(., file = ground_output)
structural_data %>%
  write_csv(., file = structural_output)
shrub_data %>%
  write_csv(., file = shrub_output)
environment_data %>%
  write_csv(., file = environment_output)
soilmetrics_data %>%
  write_csv(., file = soilmetrics_output)
soilhorizons_data %>%
  write_csv(., file = soilhorizons_output)
